teos/cli/teos_cli.py

```diff
@@ -110,85 +155,267 @@ def stop(self):
+    Attributes:
+        rpc_client (:class:`RpcClient`): the rpc client that is passed to the ``run`` method of the commands.
+    """
 
-                if command == "get_all_appointments":
-                    sys.exit(help_get_all_appointments())
+    # A dictionary mapping each command's name to the corresponding CliCommand subclass.
+    # It is populated by the ``_register_command`` decorator.
+    COMMANDS = {}
 
-                elif command == "get_tower_info":
-                    sys.exit(help_get_tower_info())
+    def __init__(self, data_dir, command_line_conf):
+        # Loads config and sets up the data folder and log file
+        config_loader = ConfigLoader(data_dir, CONF_FILE_NAME, DEFAULT_CONF, command_line_conf)
+        config = config_loader.build_config()
 
-                elif command == "get_users":
-                    sys.exit(help_get_users())
+        setup_data_folder(data_dir)
 
-                elif command == "get_user":
-                    sys.exit(help_get_user())
+        teos_rpc_host = config.get("RPC_BIND")
+        teos_rpc_port = config.get("RPC_PORT")
 
-                elif command == "stop":
-                    sys.exit(help_stop())
+        self.rpc_client = RPCClient(teos_rpc_host, teos_rpc_port)
 
-                else:
-                    sys.exit("Unknown command. Use help to check the list of available commands")
+    @classmethod
+    def _register_command(cls, command_cls):
+        """
+        Register a new command, which must be a subclass of :class:`CliCommand` and must override the ``name`` field
+        with an appropriate string.
 
+        Raises:
+            :obj:`TypeError`: ``command_cls`` is not a subclass of :class:`CliCommand`, or its ``name`` field is not a
+            string.
+        """
+
+        if not issubclass(command_cls, CliCommand):
+            raise TypeError(f"{command_cls.__name__} is not a subclass of CliCommand")
+
+        if not isinstance(command_cls.name, str):
+            raise TypeError(f'The "name" attribute of {command_cls.__name__} must be a string.')
+
+        cls.COMMANDS[command_cls.name] = command_cls
+        return command_cls
+
+    def run(self, command_name, raw_args):
+        """
+        Parses ``raw_args`` using the ``parse_args`` method of the command.
+        Then, executes the command's ``run`` method, passing the ``rpc_client`` and the output of ``parse_args``.
+        It any error that might happen, showing an appropriate message to console.
+
+        Returns:
+            The return value of the ``run`` command, or :obj:`None` if an error was raised.
+        """
+
+        if command_name not in self.COMMANDS:
+            sys.exit("Unknown command. Use help to check the list of available commands")
+
+        cmd = self.COMMANDS[command_name]
+
+        try:
+            args = cmd.parse_args(raw_args)
+            return self.COMMANDS[command_name].run(self.rpc_client, args)
+        except grpc.RpcError as e:
```

Comment:
**Check global comment first**

If we create a custom exception we'll be able to remove grpc from here.

Reply:
I moved RPCClient to a separate file, but didn't create a custom exception yet; I don't quite see a large benefit in doing so, and it would mean wrapping every call from the RPCClient to handle RpcError and reraise the custom exception.